### PR TITLE
Move integration tests into enarx-keep

### PR DESCRIPTION
**Pull request by lkatalin**
To do: test `read()`

### CODE REVIEW

File: enarx-keep/build.rs [hunk 1]
```diff
@@ -30,21 +33,78 @@ fn rerun_src(path: impl AsRef<Path>) -> std::io::Result<()> {
...
+    create_dir(&asm_out_dir);
+
+    for entry in path.read_dir().expect("failed to read test-bin dir") {
+        let file = entry.expect("failed to read file in test-bin dir");
+        let f = file.file_name();
+        let filename = f.to_str().unwrap();
+        let output: &str = filename.split('.').collect::<Vec<&str>>()[0];
+
+        std::process::Command::new(&compiler)
+            .current_dir(&asm_out_dir)
```
Comment (haraldh) [inline enarx-keep/build.rs:85]:
```suggestion
        let mut cmd = cc::Build::new()
            .no_default_flags(true)
            .get_compiler()
            .to_command();

        cmd.current_dir(&asm_out_dir)
```

**Review by haraldh — CHANGES_REQUESTED** [review]

File: enarx-keep/src/backend/kvm/vm/x86_64.rs [hunk 2]
```diff
@@ -15,6 +15,7 @@ pub struct VMSetup<'a> {
 }
 
 impl VMSetup<'_> {
+    #[allow(dead_code)]
```
Comment (haraldh) [inline enarx-keep/src/backend/kvm/vm/x86_64.rs:18]: please don't ... let @connorkuehl decide

**Comment by haraldh** [comment]
Needs https://github.com/enarx/enarx/pull/885 before the CI succeeds on KVM/SEV

File: enarx-keep/src/backend/kvm/vm/x86_64.rs (hunk 2 above)
Comment (lkatalin) [reply enarx-keep/src/backend/kvm/vm/x86_64.rs:18]: Wouldn't dream of editing the `kvm` crate normally, but it was emitting a warning that was showing up in my `stderr` test. Maybe there is another way to keep from picking that up.

Comment (haraldh) [reply enarx-keep/src/backend/kvm/vm/x86_64.rs:18]: How so? Is it recompiling everything?

Comment (connorkuehl) [reply enarx-keep/src/backend/kvm/vm/x86_64.rs:18]: My opinion here is that we should just remove the dead code rather than `#[allow(dead_code)]` -- this function was superseded by one of Harald's PRs anyway so it's no longer required.

I'm also surprised to hear that the compilation process of the crate is polluting the integration tests stderr. I think that should be addressed directly.

Comment (lkatalin) [reply enarx-keep/src/backend/kvm/vm/x86_64.rs:18]: I was having a lot of trouble with extraneous cargo output in the `stderr` test generally, and this happened pretty late in the day, so between those two things it's likely that the error is on my end. I'll remove this change and see if it still crops up.

File: enarx-keep/build.rs (hunk 1 above)
Comment (lkatalin) [reply enarx-keep/build.rs]: Thanks @haraldh , that seemed to work.

File: enarx-keep/build.rs [hunk 3]
```diff
@@ -30,21 +33,56 @@ fn rerun_src(path: impl AsRef<Path>) -> std::io::Result<()> {
...
+fn create_dir(path: &PathBuf) {
+    match std::fs::create_dir(&path) {
+        Err(e) if e.kind() == std::io::ErrorKind::AlreadyExists => {}
+        Err(e) => {
+            eprintln!("Can't create {:#?} : {:#?}", path, e);
+            std::process::exit(1);
+        }
+        Ok(_) => {}
+    }
+}
```
Comment (haraldh) [inline enarx-keep/build.rs:45]: might be sufficient to call `std::fs::create_dir_all()` instead of `create_dir()`

File: enarx-keep/build.rs [hunk 4]
```diff
@@ -30,21 +33,56 @@ fn rerun_src(path: impl AsRef<Path>) -> std::io::Result<()> {
...
+
+        cmd.current_dir(&asm_out_dir)
+            .arg("-nostdlib")
+            .arg("-static-pie")
+            .arg("-fPIC")
+            .arg("-o")
+            .arg(output)
+            .arg(test_dir.join(filename))
+            .status()
+            .expect("failed to compile binary");
```
Comment (haraldh) [inline enarx-keep/build.rs:71]: you need to check for the status also, as it can be a return value != 0

Comment (haraldh) [reply enarx-keep/build.rs:71]:
```rust
    if !status.success() {
        panic!("Failed to compile {}", filename);
    }
```

File: enarx-keep/tests/integration_tests.rs [hunk 5]
```diff
@@ -0,0 +1,173 @@
...
+        .current_dir(CRATE)
+        .arg("+nightly")
+        .env_clear()
+        .envs(filtered_env)
+        .arg("run")
+        .arg("-q")
+        .arg("exec")
+        .arg(bin)
+        .spawn()
+        .expect("failed to run test bin");
```
Comment (npmccallum) [inline enarx-keep/tests/integration_tests.rs:77]: This is why you're having so much problem with extra output and having to work around it with a bunch of hacks. When `cargo test` is run, the binary is already built before the tests are run. You just need to execute it directly rather than using `cargo +nightly run`.

I created a [working example of this strategy](https://github.com/npmccallum/demo/). We execute the binary [like this](https://github.com/npmccallum/demo/blob/master/tests/test.rs#L7-L8). This gives no extra output we have to work around.

In short, if you run the binary directly and don't use `cargo run` you can probably remove half the code in this commit which works around the problems of using `cargo run`.

**Comment by npmccallum** [comment]
One other thing I forgot: your `.s` files have mixed tabs and spaces.

File: enarx-keep/tests/test-bin/clock_gettime.s [hunk 6]
```diff
@@ -0,0 +1,17 @@
...
+ 	lea ts(%rip),   %rsi /* timespec */
+        syscall
+
+        mov $60,   	%rax /* SYS_exit */
+        mov $0,    	%rdi /* exit status */
+        syscall
+
+.data
+ts:
+	.space 128
```
[the remark this answers is not in the document]
Comment (lkatalin) [reply enarx-keep/tests/test-bin/clock_gettime.s:17]: @npmccallum Why is it preferable to put this on the stack?

File: enarx-keep/build.rs [hunk 7]
```diff
@@ -30,21 +33,56 @@ fn rerun_src(path: impl AsRef<Path>) -> std::io::Result<()> {
     Ok(())
 }
 
+fn create_dir(path: &PathBuf) {
+    match std::fs::create_dir(&path) {
+        Err(e) if e.kind() == std::io::ErrorKind::AlreadyExists => {}
+        Err(e) => {
+            eprintln!("Can't create {:#?} : {:#?}", path, e);
+            std::process::exit(1);
+        }
```
[the remark this answers is not in the document]
Comment (lkatalin) [reply enarx-keep/build.rs:42]: @npmccallum I changed the code to use `std::fs::create_dir_all()` instead of this separate function based on @haraldh 's suggestion, since the former will not error if a directory already exists. Is there a reason we should *not* do it that way?

File: enarx-keep/tests/integration_tests.rs (hunk 5 above)
Comment (lkatalin) [reply enarx-keep/tests/integration_tests.rs:30]: @npmccallum Thank you!!!

**Comment by lkatalin** [comment]
This new version ended up not needing @haraldh 's code since @npmccallum pointed out the better way to run the test is not using `cargo`.

File: enarx-keep/build.rs (hunk 7 above)
[the remark this answers is not in the document]
Comment (lkatalin) [reply enarx-keep/build.rs]: Just kidding, I ended up leaving the original code as-is.

File: enarx-keep/tests/bin/clock_gettime.c [hunk 8]
```diff
@@ -0,0 +1,21 @@
+#include <sys/syscall.h>
+#include <time.h>
+
+struct timespec t;
+
+void _start(void) {
```
Comment (haraldh) [inline enarx-keep/tests/bin/clock_gettime.c:6]:
```suggestion
__attribute__((naked)) void _start(void) {
```

gives better generated code with `ud2` after the last syscall:

```asm
0000000000001000 <_start>:
    1000:	ba e4 00 00 00       	mov    $0xe4,%edx
    1005:	b9 01 00 00 00       	mov    $0x1,%ecx
    100a:	48 8d 05 ef 2f 00 00 	lea    0x2fef(%rip),%rax        # 4000 <t>
    1011:	48 89 c6             	mov    %rax,%rsi
    1014:	89 d0                	mov    %edx,%eax
    1016:	89 cf                	mov    %ecx,%edi
    1018:	0f 05                	syscall 
    101a:	b8 3c 00 00 00       	mov    $0x3c,%eax
    101f:	ba 00 00 00 00       	mov    $0x0,%edx
    1024:	89 d7                	mov    %edx,%edi
    1026:	0f 05                	syscall 
    1028:	90                   	nop
    1029:	0f 0b                	ud2    
```

File: enarx-keep/tests/bin/exit_one.c [hunk 9]
```diff
@@ -0,0 +1,12 @@
+#include <sys/syscall.h>
+
+#define STATUS 1
+
+void _start(void) {
```
Comment (haraldh) [inline enarx-keep/tests/bin/exit_one.c:5]:
```suggestion
__attribute__((naked)) void _start(void) {
```

File: enarx-keep/tests/bin/exit_zero.c [hunk 10]
```diff
@@ -0,0 +1,12 @@
+#include <sys/syscall.h>
+
+#define STATUS 0
+
+void _start(void) {
```
Comment (haraldh) [inline enarx-keep/tests/bin/exit_zero.c:5]:
```suggestion
__attribute__((naked)) void _start(void) {
```

File: enarx-keep/tests/bin/write_stderr.c [hunk 11]
```diff
@@ -0,0 +1,22 @@
+#include <sys/syscall.h>
+#include <unistd.h>
+
+void _start(void) {
```
Comment (haraldh) [inline enarx-keep/tests/bin/write_stderr.c:4]:
```suggestion
__attribute__((naked)) void _start(void) {
```

File: enarx-keep/tests/bin/write_stdout.c [hunk 12]
```diff
@@ -0,0 +1,22 @@
+#include <sys/syscall.h>
+#include <unistd.h>
+
+void _start(void) {
```
Comment (haraldh) [inline enarx-keep/tests/bin/write_stdout.c:4]:
```suggestion
__attribute__((naked)) void _start(void) {
```

**Review by haraldh — CHANGES_REQUESTED** [review]

File: enarx-keep/tests/bin/clock_gettime.c (hunk 8 above)
Comment (lkatalin) [reply enarx-keep/tests/bin/clock_gettime.c]: Thanks @haraldh , just updated this. What does the `ud2` get us? Why do we want an invalid opcode after the last syscall?

Comment (lkatalin) [reply enarx-keep/tests/bin/clock_gettime.c]: @haraldh Looks like the naked function does not want to work with the strings:  `cannot allocate stack for variable s, naked function`.

Comment (haraldh) [reply enarx-keep/tests/bin/clock_gettime.c]: > @haraldh Looks like the naked function does not want to work with the strings: `cannot allocate stack for variable s, naked function`.

@lkatalin boooh.. bad compiler.. bad suggestion of me :-1:

Comment (haraldh) [reply enarx-keep/tests/bin/clock_gettime.c]: just add `ud2` manually

Comment (npmccallum) [reply enarx-keep/tests/bin/clock_gettime.c]: @lkatalin `ud2` always causes a CPU fault. Basically, it is the manual way of guaranteeing that the `exit` syscall never returns without a CPU fault.

Comment (lkatalin) [reply enarx-keep/tests/bin/clock_gettime.c:6]: Makes sense now!

**Review by haraldh — APPROVED** [review]

File: enarx-keep/tests/integration_tests.rs [hunk 13]
```diff
@@ -0,0 +1,103 @@
...
+        return status.success();
+    } else {
+        cmd.kill().unwrap();
+        panic!("error: test timeout");
+    }
+}
+
+/// Returns a handle to a child process through which output (stdout, stderr) can
+/// be accessed.
+fn run_test(bin: PathBuf) -> std::process::Child {
```
Comment (npmccallum) [inline enarx-keep/tests/integration_tests.rs:42]:
```suggestion
fn run_test(bin: &str, code: Option<i32>) -> std::process::Child {
```

See below...

File: enarx-keep/tests/integration_tests.rs [hunk 14]
```diff
@@ -0,0 +1,103 @@
...
+        .stderr(Stdio::piped())
+        .spawn()
+        .expect("failed to run test bin");
+
+    if let Ok(None) = cmd.wait_timeout(Duration::from_secs(TIMEOUT_SECS)) {
+        cmd.kill().unwrap();
+        panic!("error: test timeout");
+    } else {
+        cmd
+    }
```
Comment (npmccallum) [inline enarx-keep/tests/integration_tests.rs:58]:
```suggestion
    }
    
    assert_eq!(cmd.code(), code);
    cmd
```

File: enarx-keep/tests/integration_tests.rs [hunk 15]
```diff
@@ -0,0 +1,103 @@
...
+        .spawn()
+        .expect("failed to run test bin");
+
+    if let Some(status) = cmd.wait_timeout(Duration::from_secs(TIMEOUT_SECS)).unwrap() {
+        return status.success();
+    } else {
+        cmd.kill().unwrap();
+        panic!("error: test timeout");
+    }
+}
```
Comment (npmccallum) [inline enarx-keep/tests/integration_tests.rs:38]: Remove this whole function.

File: enarx-keep/tests/integration_tests.rs [hunk 16]
```diff
@@ -0,0 +1,103 @@
...
+const OUT_DIR: &str = env!("OUT_DIR");
+const TEST_BINS_OUT: &str = "bin";
+const TIMEOUT_SECS: u64 = 5;
+
+fn path(input: &str) -> PathBuf {
+    Path::new(CRATE)
+        .join(OUT_DIR)
+        .join(TEST_BINS_OUT)
+        .join(input)
+}
```
Comment (npmccallum) [inline enarx-keep/tests/integration_tests.rs:20]: Merge this function into `run_test()` below.

File: enarx-keep/tests/integration_tests.rs [hunk 17]
```diff
@@ -0,0 +1,103 @@
...
+        panic!("error: test timeout");
+    } else {
+        cmd
+    }
+}
+
+#[test]
+fn exit_zero() {
+    let code = path("exit_zero");
+    assert!(run_succeeds(code));
```
Comment (npmccallum) [inline enarx-keep/tests/integration_tests.rs:64]:
```suggestion
    run_test("exit_zero", Some(0));
```

Comment (npmccallum) [reply enarx-keep/tests/integration_tests.rs:64]: ... and so forth for all tests.

File: enarx-keep/tests/integration_tests.rs [hunk 18]
```diff
@@ -0,0 +1,82 @@
+// SPDX-License-Identifier: Apache-2.0
+
+use std::io::Read;
+use std::path::Path;
+use std::process::{Command, Stdio};
+use std::time::Duration;
+use wait_timeout::ChildExt;
+
+const CRATE: &str = env!("CARGO_MANIFEST_DIR");
+const KEEP_BIN: &str = "/target/debug/enarx-keep";
```
Comment (npmccallum) [inline enarx-keep/tests/integration_tests.rs:10]: This will break the moment we do `cargo test --profile=release`.

File: enarx-keep/tests/integration_tests.rs [hunk 19]
```diff
@@ -0,0 +1,82 @@
...
+const OUT_DIR: &str = env!("OUT_DIR");
+const TEST_BINS_OUT: &str = "bin";
+const TIMEOUT_SECS: u64 = 5;
+
+/// Returns a handle to a child process through which output (stdout, stderr) can
+/// be accessed.
+fn run_test(bin: &str, status: Option<i32>) -> std::process::Child {
```
Comment (npmccallum) [inline enarx-keep/tests/integration_tests.rs:17]: Since we always specify an exit status...

```suggestion
fn run_test(bin: &str, status: i32) -> std::process::Child {
```

**Review by haraldh — APPROVED** [review]
